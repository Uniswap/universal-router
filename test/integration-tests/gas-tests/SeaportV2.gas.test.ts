import { CommandType, RoutePlanner } from '../shared/planner'
import { UniversalRouter, Permit2 } from '../../../typechain'
import snapshotGasCost from '@uniswap/snapshot-gas-cost'
import {
  seaportV2Orders,
  seaportV2Interface,
  getAdvancedOrderParams,
  ZERO_CONDUIT_KEY,
} from '../shared/protocolHelpers/seaport'
import { resetFork } from '../shared/mainnetForkHelpers'
import { ALICE_ADDRESS, DEADLINE } from '../shared/constants'
import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers'
import hre from 'hardhat'
import deployUniversalRouter, { deployPermit2 } from '../shared/deployUniversalRouter'
const { ethers } = hre

describe('SeaportV2 Gas Tests', () => {
  let alice: SignerWithAddress
  let router: UniversalRouter
  let permit2: Permit2
  let planner: RoutePlanner

  describe('ETH -> NFT', () => {
    beforeEach(async () => {
      await resetFork(16592843 - 1) // 1 block before the order was created
      await hre.network.provider.request({
        method: 'hardhat_impersonateAccount',
        params: [ALICE_ADDRESS],
      })
      alice = await ethers.getSigner(ALICE_ADDRESS)
      permit2 = (await deployPermit2()).connect(alice) as Permit2
      router = (await deployUniversalRouter(permit2)).connect(alice) as UniversalRouter
      planner = new RoutePlanner()
    })

    it('gas: fulfillAdvancedOrder', async () => {
      const { advancedOrder, value } = getAdvancedOrderParams(seaportV2Orders[0])
      const calldata = seaportV2Interface.encodeFunctionData('fulfillAdvancedOrder', [
        advancedOrder,
        [],
        ZERO_CONDUIT_KEY,
        alice.address,
      ])

      planner.addCommand(CommandType.SEAPORT_V2, [value.toString(), calldata])
      const { commands, inputs } = planner

      await snapshotGasCost(router['execute(bytes,bytes[],uint256)'](commands, inputs, DEADLINE, { value }))
    })
  })
})
