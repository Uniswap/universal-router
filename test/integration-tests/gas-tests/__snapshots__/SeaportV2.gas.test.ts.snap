// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeaportV2 Gas Tests ETH -> NFT gas: fulfillAdvancedOrder 1`] = `
Object {
  "calldataByteLength": 2052,
  "gasUsed": 172055,
}
`;
